diff --git a/source/discovery/carla-discovery.cpp b/source/discovery/carla-discovery.cpp
index 1be334e45..b496de35a 100644
--- a/source/discovery/carla-discovery.cpp
+++ b/source/discovery/carla-discovery.cpp
@@ -69,6 +69,8 @@
 #ifdef CARLA_OS_WIN
 # include <pthread.h>
 # include <objbase.h>
+# include <iostream>
+# include <fstream>
 #endif
 
 #ifdef HAVE_FLUIDSYNTH
@@ -1668,6 +1670,21 @@ int main(int argc, char* argv[])
     CarlaString filenameCheck(filename);
     filenameCheck.toLower();
 
+    CarlaString pluginPath(filename);
+    CarlaString discoveryPath(argv[0]);
+
+#ifdef CARLA_OS_WIN64
+    CarlaString discoveryBit("64");
+#else
+    CarlaString discoveryBit("32");
+#endif
+  
+    char outPath[MAX_PATH];
+    sprintf(outPath, "%s\\carla-discovery_%s_%s", getenv("TEMP"), discoveryBit.buffer(), pluginPath.toBasic().buffer());
+
+    std::ofstream buf(outPath);
+    std::cout.rdbuf(buf.rdbuf());
+
     bool openLib = false;
     lib_t handle = nullptr;
 
@@ -1856,6 +1873,7 @@ int main(int argc, char* argv[])
     if (openLib && handle != nullptr)
         lib_close(handle);
 
+
     // ---------------------------------------------------------------------------------------------------------------
 
 #ifdef CARLA_OS_WIN
@@ -1867,6 +1885,7 @@ int main(int argc, char* argv[])
     OleUninitialize();
 #endif
 
+    buf.close();
     return 0;
 
 #ifdef USING_JUCE
