diff --git a/source/discovery/carla-discovery.cpp b/source/discovery/carla-discovery.cpp
index d45e6a12a..1684c1ed4 100644
--- a/source/discovery/carla-discovery.cpp
+++ b/source/discovery/carla-discovery.cpp
@@ -68,6 +68,8 @@
 #ifdef CARLA_OS_WIN
 # include <pthread.h>
 # include <objbase.h>
+# include <iostream>
+# include <fstream>
 #endif
 
 #ifdef HAVE_FLUIDSYNTH
@@ -1652,6 +1654,15 @@ int main(int argc, char* argv[])
         return 1;
     }
 
+
+    CarlaString out(argv[2]);
+    
+    char outPath[MAX_PATH];
+    sprintf(outPath, "%s/carla-discovery_%s", getenv("TEMP"), out.toBasic().toLower().buffer());
+    
+    std::ofstream buf(outPath);
+    std::cout.rdbuf(buf.rdbuf());
+
     const char* const stype    = argv[1];
     const char* const filename = argv[2];
     const PluginType  type     = getPluginTypeFromString(stype);
